lib/beautiful/gtk.lua

```diff
@@ -0,0 +1,282 @@
+local gears_debug = require("gears.debug")
+local gears_math = require("gears.math")
+local join = require("gears.table").join
+local unpack = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
+
+
+local gtk = {
+    cached_theme_variables = nil
+}
+
+
+local function convert_gtk_channel_to_hex(channel_value)
+    return string.format("%02x", gears_math.round(channel_value * 255))
+end
+
+local function convert_gtk_color_to_hex(gtk_color)
+    return "#" ..
+        convert_gtk_channel_to_hex(gtk_color.red) ..
+        convert_gtk_channel_to_hex(gtk_color.green) ..
+        convert_gtk_channel_to_hex(gtk_color.blue) ..
+        convert_gtk_channel_to_hex(gtk_color.alpha)
+end
+
+local function lookup_gtk_color_to_hex(_style_context, color_name)
+    local gtk_color = _style_context:lookup_color(color_name)
+    if not gtk_color then
+        return nil
+    end
+    return convert_gtk_color_to_hex(gtk_color)
+end
+
+local function get_gtk_property(_style_context, property_name)
+    local state = _style_context:get_state()
+    local property = _style_context:get_property(property_name, state)
+    if not property then
+        return nil
+    end
+    return property.value
+end
+
+local function get_gtk_color_property_to_hex(_style_context, property_name)
+    return convert_gtk_color_to_hex(
+        get_gtk_property(_style_context, property_name)
+    )
+end
+
+local function read_gtk_color_properties_from_widget(gtk_widget, properties)
+    local _style_context = gtk_widget:get_style_context()
+    local result = {}
+    for result_key, style_context_property in pairs(properties) do
+        result[result_key] = get_gtk_color_property_to_hex(
+            _style_context, style_context_property
+        )
+    end
+    return result
+end
+
+
+-- luacheck: max comment line length 300
+
+--- Get GTK+3 theme variables from GtkStyleContext
+-- @treturn table Key-value table with the following structure:
+-- <table class='widget_list' border=1>
+-- <tr style='font-weight: bold;'> <th align='center'>Result key</th> <th align='center'>StyleContext key</th> <th align='center'>StyleContext fallback #1</th> <th align='center'>StyleContext fallback #2</th> <th align='center'>GTK Widget fallback</th> </tr>
+-- <tr> <td>`font_size`</td> <td></td> <td></td> <td></td> <td>Label font-size</td> </tr>
+-- <tr> <td>`font_family`</td> <td></td> <td></td> <td></td> <td>Label font-family</td> </tr>
+-- <tr> <td>`bg_color`</td> <td>`theme_bg_color`</td> <td></td> <td></td> <td>Window bg</td> </tr>
+-- <tr> <td>`fg_color`</td> <td>`theme_fg_color`</td> <td></td> <td></td> <td>Window fg</td> </tr>
+-- <tr> <td>`base_color`</td> <td>`theme_base_color`</td> <td></td> <td></td> <td>Entry bg</td> </tr>
+-- <tr> <td>`text_color`</td> <td>`theme_text_color`</td> <td></td> <td></td> <td>Entry fg</td> </tr>
+-- <tr> <td>`button_bg_color`</td> <td>`theme_buton_bg_color`</td> <td>`theme_bg_color`</td> <td></td> <td>Button bg</td> </tr>
```

Comment:
There's a small typo in `theme_buton_bg_color`.

Reply:
thanks for your eagle eye!